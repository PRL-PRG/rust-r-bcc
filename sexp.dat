A
3
263168
197888
5
UTF-8
6
1
262153
1
+
2
1
262153
1
x
2
14
1
1
254
