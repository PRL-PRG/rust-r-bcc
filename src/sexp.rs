use std::rc::Rc;

pub struct Loc {
    row: usize,
    col: usize,
}

#[derive(Debug, PartialEq)]
pub struct MetaData;

#[derive(Debug, PartialEq)]
pub struct Sexp {
    kind: SexpKind,
    metadata: MetaData,
}

impl From<SexpKind> for Sexp {
    fn from(kind: SexpKind) -> Self {
        Sexp {
            kind,
            metadata: MetaData,
        }
    }
}

pub mod data {
    use std::rc::Rc;

    use super::{Sexp, SexpKind};

    #[derive(Debug, PartialEq)]
    pub struct TaggedSexp {
        // if this will be imutable then
        // Rc will be better for coping
        tag: Option<Rc<str>>,
        data: Sexp,
    }

    impl TaggedSexp {
        pub fn new(data: Sexp) -> Self {
            Self { tag: None, data }
        }
    }

    impl From<Sexp> for TaggedSexp {
        fn from(value: Sexp) -> Self {
            Self::new(value)
        }
    }

    impl From<SexpKind> for TaggedSexp {
        fn from(value: SexpKind) -> Self {
            Self::new(value.into())
        }
    }

    #[derive(Debug, PartialEq)]
    pub struct Complex {
        real: f64,
        imaginary: f64,
    }

    pub type List = Vec<TaggedSexp>;
}

pub mod lang {
    use std::rc::Rc;

    #[derive(Debug, PartialEq)]
    pub struct Sym {
        data: Rc<str>,
    }

    #[derive(Debug, PartialEq)]
    pub enum Target {}

    #[derive(Debug, PartialEq)]
    pub struct Lang {
        target: Target,
        args: super::data::List,
    }
}

// SXP
#[derive(Debug, PartialEq)]
pub enum SexpKind {
    Sym(lang::Sym),
    List(data::List),
    Nil,

    // language contructs
    Closure,
    Environment,
    Promise,
    Lang(lang::Lang),

    // vecs
    Char(Vec<char>),
    Logic(Vec<bool>),
    Real(Vec<f64>),
    Int(Vec<i32>),
    Complex(Vec<data::Complex>),
    Str(Vec<Rc<str>>),
    Vec(Vec<Sexp>),
}
