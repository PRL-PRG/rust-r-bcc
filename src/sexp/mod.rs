pub mod bc;
mod bc_display;
pub mod sexp;
pub mod sexp_alloc;
mod sexp_display;
