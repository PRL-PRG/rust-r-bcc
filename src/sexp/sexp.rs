use super::bc::Bc;

pub struct Loc {
    row: usize,
    col: usize,
}

#[derive(Debug, PartialEq, Clone, Default)]
pub struct MetaData {
    pub attr: Option<Box<Sexp>>,
}

impl MetaData {
    pub fn is_obj(&self) -> bool {
        if self.attr.is_some() {
            let attr = self.attr.clone().unwrap();
            if let Sexp {
                kind: SexpKind::List(list),
                ..
            } = *attr
            {
                list.into_iter().any(|x| {
                    let x = match x.tag {
                        Some(x) if x.as_str() == "class" => true,
                        _ => false,
                    };
                    x
                })
            } else {
                false
            }
        } else {
            false
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Sexp {
    pub kind: SexpKind,
    pub metadata: MetaData,
}

impl From<SexpKind> for Sexp {
    fn from(kind: SexpKind) -> Self {
        Sexp {
            kind,
            metadata: MetaData { attr: None },
        }
    }
}

impl Sexp {
    pub fn set_attr(&mut self, attr: Sexp) {
        self.metadata.attr = Some(Box::new(attr))
    }
}

pub mod data {
    use super::{Sexp, SexpKind};

    #[derive(Debug, PartialEq, Clone)]
    pub struct TaggedSexp {
        pub tag: Option<String>,
        pub data: Sexp,
    }

    impl TaggedSexp {
        pub fn new(data: Sexp) -> Self {
            Self { tag: None, data }
        }

        pub fn new_with_tag(data: Sexp, tag: String) -> Self {
            let tag = Some(tag);
            Self { tag, data }
        }
    }

    impl From<Sexp> for TaggedSexp {
        fn from(value: Sexp) -> Self {
            Self::new(value)
        }
    }

    impl From<SexpKind> for TaggedSexp {
        fn from(value: SexpKind) -> Self {
            Self::new(value.into())
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    pub struct Complex {
        pub real: f64,
        pub imaginary: f64,
    }

    pub type List = Vec<TaggedSexp>;
}

pub mod lang {

    #[derive(Debug, PartialEq, Clone)]
    pub struct Sym {
        pub(crate) data: String,
    }

    impl Sym {
        pub fn new(data: String) -> Self {
            Self { data }
        }
    }

    impl From<&str> for Sym {
        fn from(value: &str) -> Self {
            Sym {
                data: value.to_string(),
            }
        }
    }

    impl Into<super::Sexp> for Sym {
        fn into(self) -> super::Sexp {
            super::SexpKind::Sym(self).into()
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    pub enum Target {
        Lang(Box<Lang>), // expression
        Sym(Sym),        // named
    }

    impl From<Sym> for Target {
        fn from(value: Sym) -> Self {
            Target::Sym(value)
        }
    }

    impl From<Lang> for Target {
        fn from(value: Lang) -> Self {
            Target::Lang(Box::new(value))
        }
    }

    impl Into<super::Sexp> for Target {
        fn into(self) -> super::Sexp {
            match self {
                Target::Lang(x) => super::SexpKind::Lang(*x).into(),
                Target::Sym(x) => super::SexpKind::Sym(x).into(),
            }
        }
    }

    impl Into<super::Sexp> for &Target {
        fn into(self) -> super::Sexp {
            match self {
                Target::Lang(x) => super::SexpKind::Lang(*x.clone()).into(),
                Target::Sym(x) => super::SexpKind::Sym(x.clone()).into(),
            }
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    pub struct Lang {
        pub(crate) target: Target,
        pub(crate) args: super::data::List,
    }

    impl Into<super::Sexp> for Lang {
        fn into(self) -> super::Sexp {
            super::SexpKind::Lang(self).into()
        }
    }

    impl Lang {
        pub fn new(target: Target, args: super::data::List) -> Self {
            Self { target, args }
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    pub struct Formal {
        pub name: Sym,
        pub value: Box<super::Sexp>,
    }

    impl Formal {
        pub fn new(name: Sym, value: super::Sexp) -> Self {
            Self {
                name,
                value: Box::new(value),
            }
        }
    }

    impl TryInto<Formal> for super::data::TaggedSexp {
        type Error = crate::rds::rds_reader::RDSReaderError;

        fn try_into(self) -> Result<Formal, Self::Error> {
            match self.tag {
                Some(name) => Ok(Formal::new(Sym::new(name), self.data)),
                None => Err(crate::rds::rds_reader::RDSReaderError::DataError(
                    "Formal must have name as a tag".to_string(),
                )),
            }
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    pub struct Closure {
        pub(crate) formals: Vec<Formal>,
        pub(crate) body: Box<super::Sexp>,
        pub(crate) environment: Environment,
    }

    impl Closure {
        pub fn new(formals: Vec<Formal>, body: super::Sexp, environment: Environment) -> Self {
            Self {
                formals,
                body: Box::new(body),
                environment,
            }
        }
    }

    impl Into<super::Sexp> for Closure {
        fn into(self) -> super::Sexp {
            super::SexpKind::Closure(self).into()
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    pub enum Environment {
        Global,
        Base,
        Empty,
        Normal(NormalEnv),
    }

    #[derive(Debug, PartialEq, Clone)]
    pub struct NormalEnv {
        pub parent: Box<Environment>,
        pub locked: bool,
        pub frame: ListFrame,
        pub hash_frame: HashFrame,
    }

    impl NormalEnv {
        pub fn new(
            parent: Box<Environment>,
            locked: bool,
            frame: ListFrame,
            hash_frame: HashFrame,
        ) -> Self {
            Self {
                parent,
                locked,
                frame,
                hash_frame,
            }
        }
    }

    impl Into<Environment> for NormalEnv {
        fn into(self) -> Environment {
            Environment::Normal(self)
        }
    }

    impl Into<super::Sexp> for Environment {
        fn into(self) -> super::Sexp {
            super::SexpKind::Environment(self).into()
        }
    }

    #[derive(Debug, PartialEq, Clone, Default)]
    pub struct ListFrame {
        pub data: Option<super::data::List>,
    }

    impl ListFrame {
        pub fn new(data: super::data::List) -> Self {
            Self { data: Some(data) }
        }
    }

    #[derive(Debug, PartialEq, Clone, Default)]
    pub struct HashFrame {
        pub data: Option<Vec<super::Sexp>>,
    }

    impl HashFrame {
        pub fn new(data: Vec<super::Sexp>) -> Self {
            Self { data: Some(data) }
        }
    }

    impl Into<super::SexpKind> for Environment {
        fn into(self) -> super::SexpKind {
            super::SexpKind::Environment(self)
        }
    }
}

// SXP
#[derive(Debug, PartialEq, Clone)]
pub enum SexpKind {
    Sym(lang::Sym),
    List(data::List),
    Nil,

    // language contructs
    Closure(lang::Closure),
    Environment(lang::Environment),
    Promise,
    Lang(lang::Lang),
    Bc(Bc),

    // vecs
    Char(Vec<char>),
    Logic(Vec<bool>),
    Real(Vec<f64>),
    Int(Vec<i32>),
    Complex(Vec<data::Complex>),
    Str(Vec<String>),
    Vec(Vec<Sexp>),

    MissingArg,
}
