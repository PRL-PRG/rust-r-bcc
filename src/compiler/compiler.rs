use crate::sexp::{
    bc::{Bc, BcOp},
    sexp::{data, lang, MetaData, Sexp, SexpKind},
};

#[derive(Default, Clone)]
struct CompilerContext {
    top_level: bool,
    tailcall: bool,
    call: Option<Sexp>,
}

impl CompilerContext {
    fn new_top(ctxt: &CompilerContext) -> Self {
        Self {
            top_level: true,
            tailcall: true,
            call: None,
            ..ctxt.clone()
        }
    }

    fn new_promise(ctxt: &CompilerContext) -> Self {
        Self {
            top_level: false,
            tailcall: true,
            ..ctxt.clone()
        }
    }

    fn new_call(ctxt: &CompilerContext, call: Sexp) -> Self {
        Self {
            call: Some(call.clone()),
            ..ctxt.clone()
        }
    }
}

/// i32 min represents NA in intepreter
const NA: i32 = i32::MIN;

#[derive(Debug)]
pub enum Warning {
    VariableDoesNotExist(String),
}

struct CodeBuffer {
    bc: Bc,
    current_expr: Option<Sexp>,
    expression_buffer: Vec<i32>,
}

impl CodeBuffer {
    fn new() -> Self {
        Self {
            bc: Bc::new(),
            current_expr: None,
            expression_buffer: vec![NA], // first instruction is version and does not have a source
        }
    }

    fn new_with_expr(curr_expr: Sexp) -> Self {
        Self {
            bc: Bc::new(),
            current_expr: Some(curr_expr),
            expression_buffer: vec![NA], // first instruction is version and does not have a source
        }
    }

    fn insert_currexpr(&mut self, bc_count: usize) {
        if self.current_expr.is_some() {
            let expr = self.current_expr.clone().unwrap();
            let index = self.add_const(expr.clone());
            self.expression_buffer
                .append(&mut (0..bc_count).map(|_| index).collect());
        }
    }

    fn add_instr(&mut self, op: BcOp) {
        assert_eq!(op.arity(), 0, "Wrong arity");
        self.bc.instructions.push(op.into());
        self.insert_currexpr(1);
    }

    fn add_instr2(&mut self, op: BcOp, idx: i32) {
        assert_eq!(op.arity(), 1, "Wrong arity");
        self.bc.instructions.push(op.into());
        self.bc.instructions.push(idx);
        self.insert_currexpr(2);
    }

    fn add_instr_n(&mut self, op: BcOp, idxs: &[i32]) {
        assert_eq!(op.arity(), idxs.len(), "Wrong arity");
        self.bc.instructions.push(op.into());
        self.bc.instructions.extend_from_slice(idxs);
        self.insert_currexpr(1 + idxs.len());
    }

    fn add_const(&mut self, val: Sexp) -> i32 {
        match self.bc.constpool.iter().position(|x| x == &val) {
            Some(idx) => idx as i32,
            None => {
                self.bc.constpool.push(val);
                (self.bc.constpool.len() - 1) as i32
            }
        }
    }

    fn set_current_expr(&mut self, sexp: Sexp) -> Option<Sexp> {
        println!("set curr expr {sexp}");
        std::mem::replace(&mut self.current_expr, Some(sexp))
    }

    fn restore_current_expr(&mut self, orig: Option<Sexp>) {
        let _ = std::mem::replace(&mut self.current_expr, orig);
    }
}

pub struct Compiler {
    options: CompilerOptions,
    context: CompilerContext,
    code_buffer: CodeBuffer,

    environment: lang::Environment,
    pub warnings: Vec<Warning>,
}

impl Compiler {
    pub fn new() -> Self {
        Self {
            options: CompilerOptions,
            context: CompilerContext::new_top(&CompilerContext::default()),
            code_buffer: CodeBuffer::new(),

            environment: lang::Environment::Global,
            warnings: vec![],
        }
    }

    pub fn cmpfun(&mut self, closure: lang::Closure) -> lang::Closure {
        let mut closure = closure;
        self.environment = lang::NormalEnv::new(
            Box::new(closure.environment),
            false,
            lang::ListFrame::new(
                closure
                    .formals
                    .iter()
                    .map(|x| {
                        data::TaggedSexp::new_with_tag(
                            x.value.as_ref().clone(),
                            x.name.data.clone(),
                        )
                    })
                    .collect(),
            ),
            lang::HashFrame::new(vec![]),
        )
        .into();
        let body =
            SexpKind::Bc(self.gen_code(closure.body.as_ref(), Some(closure.body.as_ref()))).into();
        closure.body = Box::new(body);
        let lang::Environment::Normal(env) = &mut self.environment else {
            unreachable!()
        };
        closure.environment = std::mem::replace(env.parent.as_mut(), lang::Environment::Global);
        closure
    }

    fn gen_code(&mut self, target: &Sexp, loc: Option<&Sexp>) -> Bc {
        let tmp = if let Some(loc) =loc {
            CodeBuffer::new_with_expr(loc.clone())
        } else {
            CodeBuffer::new() 
        };
        let orig = std::mem::replace(&mut self.code_buffer, tmp);
        self.code_buffer.add_const(target.clone());
        self.cmp(&target, false, false);
        let locs = self.create_expression_loc();
        self.code_buffer.add_const(locs);
        std::mem::replace(&mut self.code_buffer, orig).bc
    }

    fn cmp(&mut self, sexp: &Sexp, missing_ok: bool, set_loc: bool) {
        let orig = if set_loc {
            self.code_buffer.set_current_expr(sexp.clone())
        } else {
            None
        };
        match &sexp.kind {
            SexpKind::Sym(sym) => self.cmp_sym(sym, missing_ok),
            SexpKind::Nil => {
                self.code_buffer.add_instr(BcOp::LDNULL_OP);

                if self.context.tailcall {
                    self.code_buffer.add_instr(BcOp::RETURN_OP);
                }
            }
            SexpKind::Environment(_) => todo!(),
            SexpKind::Promise => todo!(),
            SexpKind::Lang(lang) => self.cmp_call(lang),
            _ => {
                let ci = self.code_buffer.add_const(sexp.clone());
                self.code_buffer.add_instr2(BcOp::LDCONST_OP, ci);

                if self.context.tailcall {
                    self.code_buffer.add_instr(BcOp::RETURN_OP);
                }
            }
        };
        self.code_buffer.restore_current_expr(orig);
    }

    fn cmp_sym(&mut self, sym: &lang::Sym, missing_ok: bool) {
        match sym.data.as_str() {
            ".." => todo!(),
            name if name.starts_with("..") => todo!(),
            name => {
                if !self.var_exist(name) {
                    self.warnings
                        .push(Warning::VariableDoesNotExist(name.to_string()))
                }
                let index = self.code_buffer.add_const(sym.clone().into());
                if missing_ok {
                    self.code_buffer.add_instr2(BcOp::GETVAR_MISSOK_OP, index);
                } else {
                    self.code_buffer.add_instr2(BcOp::GETVAR_OP, index);
                }
            }
        }
        if self.context.tailcall {
            self.code_buffer.add_instr(BcOp::RETURN_OP);
        }
    }

    fn cmp_call(&mut self, call: &lang::Lang) {
        // set up state
        let mut sexp_tmp = std::mem::take(&mut self.code_buffer.current_expr);
        let orig = self.code_buffer.set_current_expr(call.clone().into());
        let tmp = CompilerContext::new_call(&self.context, call.clone().into());
        let mut orig_context = std::mem::replace(&mut self.context, tmp);

        match &call.target {
            lang::Target::Lang(_) => todo!(),
            lang::Target::Sym(sym) => {
                let index = self.code_buffer.add_const(sym.clone().into());
                self.code_buffer.add_instr2(BcOp::GETFUN_OP, index);
                self.cmp_args(&call.args);
                let index = self.code_buffer.add_const(call.clone().into());
                self.code_buffer.add_instr2(BcOp::CALL_OP, index);
                if self.context.tailcall {
                    self.code_buffer.add_instr(BcOp::RETURN_OP);
                }
            }
        };

        // restore state
        self.code_buffer.restore_current_expr(orig);
        std::mem::swap(&mut self.context, &mut orig_context);
        std::mem::swap(&mut self.code_buffer.current_expr, &mut sexp_tmp);
    }

    fn cmp_args(&mut self, args: &data::List) {
        for arg in args {
            match &arg.data.kind {
                SexpKind::MissingArg => todo!(),
                SexpKind::Sym(sym) if sym.data.as_str() == ".." => todo!(),
                SexpKind::Bc(_) => todo!(),
                SexpKind::Promise => todo!(),
                SexpKind::Sym(_) | SexpKind::Lang(_) => {
                    let curr = self.code_buffer.current_expr.clone();
                    let code = self.gen_code(&arg.data, curr.as_ref());
                    let index = self.code_buffer.add_const(code.into());
                    self.code_buffer.add_instr2(BcOp::MAKEPROM_OP, index);
                }
                SexpKind::Nil => {
                    self.code_buffer.add_instr(BcOp::PUSHNULLARG_OP);
                    self.cmp_tag(&arg.tag);
                }
                SexpKind::Logic(logs) if logs.len() == 1 && logs[0] => {
                    self.code_buffer.add_instr(BcOp::PUSHTRUEARG_OP);
                    self.cmp_tag(&arg.tag);
                }
                SexpKind::Logic(logs) if logs.len() == 1 && !logs[0] => {
                    self.code_buffer.add_instr(BcOp::PUSHFALSEARG_OP);
                    self.cmp_tag(&arg.tag);
                }
                _ => {
                    let index = self.code_buffer.add_const(arg.data.clone());
                    self.code_buffer.add_instr2(BcOp::PUSHCONSTARG_OP, index);
                }
            }
        }
    }

    fn cmp_tag(&mut self, tag: &Option<String>) {
        if let Some(tag) = tag {
            let tag: lang::Sym = tag.as_str().into();
            let tag: Sexp = tag.into();
            let index = self.code_buffer.add_const(tag);
            self.code_buffer.add_instr2(BcOp::SETTAG_OP, index);
        }
    }

    fn var_exist(&self, name: &str) -> bool {
        self.environment.find_local_var(name).is_some()
    }

    fn create_expression_loc(&mut self) -> Sexp {
        Sexp {
            kind: SexpKind::Int(std::mem::take(&mut self.code_buffer.expression_buffer)),
            metadata: MetaData {
                attr: Some(Box::new(Sexp {
                    kind: SexpKind::List(vec![data::TaggedSexp {
                        tag: Some("class".into()),
                        data: Sexp {
                            kind: SexpKind::Str(vec!["expressionsIndex".into()]),
                            metadata: MetaData { attr: None },
                        },
                    }]),
                    metadata: MetaData { attr: None },
                })),
            },
        }
    }
}

pub struct CompilerOptions;

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::{BufWriter, Read, Write};

    use crate::rds::{rds_reader::RDSReader, rds_writer::RDSWriter, RDSResult};

    macro_rules! test_fun_noopt {
        ( $name:ident, $code:expr) => {
            mod $name {
                use super::*;

                #[test]
                fn compiler() {
                    let path = format!("temp/{}_compiler.dat", stringify!($name));
                    let path = path.as_str();
                    let path_comp = format!("temp/{}_compiler_corr.dat", stringify!($name));
                    let path_comp = path_comp.as_str();
                    let mut command = std::process::Command::new("./create_serdata.R")
                        .args(["-d", $code, path])
                        .spawn()
                        .unwrap();
                    assert!(command.wait().unwrap().success());
                    let mut command = std::process::Command::new("./create_serdata.R")
                        .args([
                            "-d",
                            format!("compiler::cmpfun({}, options=list(optimize=0))", $code)
                                .as_str(),
                            path_comp,
                        ])
                        .spawn()
                        .unwrap();
                    assert!(command.wait().unwrap().success());

                    let mut file = std::fs::File::open(path).unwrap();
                    let RDSResult {
                        header,
                        data: input,
                    } = file.read_rds().unwrap();

                    let mut input_vec = vec![];
                    let mut file = std::fs::File::open(path_comp).unwrap();
                    file.read_to_end(&mut input_vec).unwrap();

                    let mut file = std::fs::File::open(path_comp).unwrap();
                    let RDSResult {
                        header: _,
                        data: correct,
                    } = file.read_rds().unwrap();

                    println!("{}", input);

                    let mut compiler = Compiler::new();
                    let SexpKind::Closure(cl) = input.kind else {
                        unreachable!();
                    };
                    let bc = compiler.cmpfun(cl);
                    let input: Sexp = bc.into();

                    println!("My compilation:\n{input}\n");
                    println!("Correct compilation:\n{correct}");

                    let outdata: Vec<u8> = vec![];
                    let mut writer = BufWriter::new(outdata);
                    writer.write_rds(header, input).unwrap();
                    writer.flush().unwrap();

                    assert_eq!(writer.get_ref(), &input_vec);
                    std::fs::remove_file(path).unwrap();
                    std::fs::remove_file(path_comp).unwrap();
                }
            }
        };
    }

    test_fun_noopt![basic, "function() NULL"];
    test_fun_noopt![basic_real, "function() 1"];
    test_fun_noopt![identity, "function(x) x"];
    test_fun_noopt![addition, "function(x, y) x + y"];
    test_fun_noopt![block, "function(a, b = 0) {x <- c(a, 1); x[[b]];}"];
}
