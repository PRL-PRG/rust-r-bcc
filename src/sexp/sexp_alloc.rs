use std::ops::Deref;

use bumpalo::Bump;

use super::sexp::{data, lang, MetaData, Sexp, SexpKind};

pub struct Alloc<'a> {
    allocator: &'a Bump,
    pub nil: &'a Sexp<'a>,
    pub nil_list: data::List<'a>,    // this is done to dedupe empty list
    pub nil_vec: &'a [&'a Sexp<'a>], // this is done to dedupe empty vecs
    pub missing: &'a Sexp<'a>,
    pub base_env: &'a Sexp<'a>,
    pub global_env: &'a Sexp<'a>,
    pub empty_env: &'a Sexp<'a>,
    pub na_string: &'a str,
    pub unbound: &'a Sexp<'a>,

    pub empty_metadata: &'a MetaData<'a>,
}

impl<'a> Alloc<'a> {
    pub fn new(allocator: &'a Bump) -> Self {
        let nil = allocator.alloc(SexpKind::Nil.into()) as &'a Sexp;
        Self {
            allocator,
            nil,
            nil_list: data::List {
                data: allocator.alloc_slice_fill_with(0, |_| data::TaggedSexp::new(nil)),
            },
            nil_vec: allocator.alloc_slice_fill_copy(0, nil as &'a Sexp),
            missing: allocator.alloc(SexpKind::MissingArg.into()),
            base_env: allocator.alloc(SexpKind::Environment(lang::Environment::Base).into()),
            global_env: allocator.alloc(SexpKind::Environment(lang::Environment::Global).into()),
            empty_env: allocator.alloc(SexpKind::Environment(lang::Environment::Empty).into()),
            na_string: allocator.alloc_str("__NA__"),
            unbound: allocator.alloc(SexpKind::UnboundVal.into()),
            empty_metadata: allocator.alloc(MetaData::default()),
        }
    }

    // this is only for environments that are stored in
    // Alloc structure (ie. base_env)
    fn get_inner_env(env: &'a Sexp<'a>) -> &'a lang::Environment<'a> {
        let SexpKind::Environment(env) = &env.kind else {
            unreachable!()
        };

        env
    }

    pub fn get_base(&self) -> &'a lang::Environment<'a> {
        Self::get_inner_env(self.base_env)
    }

    pub fn get_global(&self) -> &'a lang::Environment<'a> {
        Self::get_inner_env(self.global_env)
    }

    pub fn get_empty(&self) -> &'a lang::Environment<'a> {
        Self::get_inner_env(self.empty_env)
    }
}

impl<'a> Deref for Alloc<'a> {
    type Target = Bump;

    fn deref(&self) -> &Self::Target {
        &self.allocator
    }
}
