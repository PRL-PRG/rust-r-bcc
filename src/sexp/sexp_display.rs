use std::fmt::Display;

use super::sexp::{data, lang, Sexp, SexpKind};

impl Display for Sexp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(args) = &self.metadata.attr {
            write!(f, "(data : {}, args : {})", self.kind, args)?;
        } else {
            write!(f, "{}", self.kind)?;
        }
        Ok(())
    }
}

pub(super) fn join_string<T>(data: &[T], sep: &str) -> String
where
    T: Display,
{
    data.iter()
        .fold(("".to_string(), true), |(acc, first), x| {
            if first {
                (x.to_string(), false)
            } else {
                (acc.to_string() + sep + x.to_string().as_str(), false)
            }
        })
        .0
}

impl Display for data::Complex {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} + {}*i", self.real, self.imaginary)
    }
}

impl Display for SexpKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SexpKind::Sym(sym) => write!(f, "{}", sym),
            SexpKind::List(data) => write!(f, "list: [{}]", join_string(data, ", ")),
            SexpKind::Nil => write!(f, "nil"),
            SexpKind::Closure(closure) => write!(f, "{closure}"),
            SexpKind::Environment(env) => write!(f, "env: ({env})"),
            SexpKind::Promise => todo!(),
            SexpKind::Lang(lang) => write!(f, "{lang}"),
            SexpKind::Char(chars) => write!(f, "'{}'", join_string(chars, "")),
            SexpKind::Logic(logs) => write!(f, "[{}]", join_string(logs, ", ")),
            SexpKind::Real(reals) => write!(f, "[{}]", join_string(reals, ", ")),
            SexpKind::Int(ints) => write!(f, "[{}]", join_string(ints, ", ")),
            SexpKind::Complex(cmplx) => write!(f, "[{}]", join_string(cmplx, ", ")),
            SexpKind::Str(strs) => write!(f, "['{}']", join_string(strs, ", ")),
            SexpKind::Vec(data) => {
                if data.is_empty() {
                    write!(f, "[]")
                } else {
                    write!(f, "{}", data[0])?;
                    for item in &data[1..] {
                        write!(f, ", {}", item)?;
                    }
                    Ok(())
                }
            }
            SexpKind::MissingArg => write!(f, "missing"),
            SexpKind::Bc(bc) => write!(f, "{bc}"),
            SexpKind::BaseNamespace => write!(f, "base namespace"),
            SexpKind::Buildin(sym) => write!(f, "{sym}"),
            SexpKind::NAString => write!(f, "NASTRING"),
        }
    }
}

impl Display for lang::Lang {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}({})", self.target, join_string(&self.args, ", "))
    }
}

impl Display for lang::Sym {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.data)
    }
}

impl Display for lang::Target {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            lang::Target::Lang(lang) => write!(f, "{lang}"),
            lang::Target::Sym(sym) => write!(f, "{}", sym),
        }
    }
}

impl Display for data::TaggedSexp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.tag {
            Some(tag) => write!(f, "({tag}) {}", self.data),
            None => write!(f, "{}", self.data),
        }
    }
}

impl Display for lang::Environment {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            lang::Environment::Global => write!(f, "global env"),
            lang::Environment::Empty => write!(f, "empty env"),
            lang::Environment::Base => write!(f, "base env"),
            lang::Environment::Normal(env) => write!(f, "{env}"),
        }
    }
}

impl Display for lang::NormalEnv {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "parent : {}", self.parent)?;
        if let Some(frame) = &self.frame.data {
            write!(f, ", frame : {}", join_string(frame, ", "))?;
        }

        if let Some(hash_frame) = &self.hash_frame.data {
            write!(f, ", hash frame : {}", join_string(hash_frame, ", "))?;
        }

        Ok(())
    }
}

impl Display for lang::Formal {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)?;

        match self.value.kind {
            SexpKind::MissingArg => (),
            _ => write!(f, " = {}", self.value)?,
        }

        Ok(())
    }
}

impl Display for lang::Closure {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "function({})\n", join_string(&self.formals, ", "))?;
        write!(f, "{}\n", self.body)?;
        write!(f, "(env : {})", self.environment)
    }
}
