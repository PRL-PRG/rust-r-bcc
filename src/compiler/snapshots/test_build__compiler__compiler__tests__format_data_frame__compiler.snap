---
source: src/compiler/compiler.rs
expression: compiler.warnings
---
[
    VariableDoesNotExist(
        "nc",
    ),
    VariableDoesNotExist(
        "nc",
    ),
    VariableDoesNotExist(
        "nc",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "NROW",
    ),
    VariableDoesNotExist(
        "lens",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "nc",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "len",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "len",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "len",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "len",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "len",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "nc",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "i",
    ),
    VariableDoesNotExist(
        "rval",
    ),
    VariableDoesNotExist(
        "nr",
    ),
    VariableDoesNotExist(
        "y",
    ),
    VariableDoesNotExist(
        "y",
    ),
]
